src/include/buffer.h

```diff
@@ -769,34 +810,44 @@ namespace buffer CEPH_BUFFER_API {
 
     // modifiers
     void clear() noexcept {
-      _buffers.clear();
+      _carriage = &always_empty_bptr;
+      _buffers.clear_and_dispose(ptr_node::disposer());
       _len = 0;
       _memcopy_count = 0;
       last_p = begin();
-      append_buffer = ptr();
     }
     void push_back(const ptr& bp) {
       if (bp.length() == 0)
 	return;
-      _buffers.push_back(bp);
+      _buffers.push_back(*ptr_node::create(bp).release());
       _len += bp.length();
     }
     void push_back(ptr&& bp) {
       if (bp.length() == 0)
 	return;
       _len += bp.length();
-      _buffers.push_back(std::move(bp));
+      _buffers.push_back(*ptr_node::create(std::move(bp)).release());
+      _carriage = &always_empty_bptr;
```

Comment:
Let's add a comment saying this is intentional as we need to take care of situations like:
```
bl.push_back(buffer::ptr(existing_one, offset, len));
```

The underlying `raw` must stay untouched.